test/QueryPlannerTest.cpp

```diff
@@ -3247,3 +3247,47 @@ TEST(QueryPlanner, testDistributiveJoinInUnionRecursive) {
                                "?_QLever_internal_variable_qp_18")))))),
       qec, {4, 16, 64'000'000});
 }
+
+// _____________________________________________________________________________
+TEST(QueryPlanner, negatedPaths) {
+  h::expect("SELECT * { ?a !<b> ?c }",
+            h::Filter("<b> != ?_QLever_internal_variable_qp_0",
+                      h::IndexScanFromStrings(
+                          "?a", "?_QLever_internal_variable_qp_0", "?c")));
+
+  h::expect("SELECT * { ?a !(<b>) ?c }",
+            h::Filter("<b> != ?_QLever_internal_variable_qp_0",
+                      h::IndexScanFromStrings(
+                          "?a", "?_QLever_internal_variable_qp_0", "?c")));
+
+  h::expect("SELECT * { ?a !^<b> ?c }",
+            h::Filter("<b> != ?_QLever_internal_variable_qp_1",
+                      h::IndexScanFromStrings(
+                          "?c", "?_QLever_internal_variable_qp_1", "?a")));
+
+  h::expect("SELECT * { ?a !(^<b>) ?c }",
+            h::Filter("<b> != ?_QLever_internal_variable_qp_1",
+                      h::IndexScanFromStrings(
+                          "?c", "?_QLever_internal_variable_qp_1", "?a")));
+  h::expect(
+      "SELECT * { ?a !(<b>|^<b>) ?c }",
+      h::Union(h::Filter(" <b> != ?_QLever_internal_variable_qp_0",
+                         h::IndexScanFromStrings(
+                             "?a", "?_QLever_internal_variable_qp_0", "?c")),
+               h::Filter("<b> != ?_QLever_internal_variable_qp_1",
+                         h::IndexScanFromStrings(
+                             "?c", "?_QLever_internal_variable_qp_1", "?a"))));
+  h::expect(
+      "SELECT * { ?a !(<b>|^<d>|^<e>|<f>) ?c }",
+      h::Union(
+          h::Filter(
+              " <f> != ?_QLever_internal_variable_qp_0",
+              h::Filter(" <b> != ?_QLever_internal_variable_qp_0",
+                        h::IndexScanFromStrings(
+                            "?a", "?_QLever_internal_variable_qp_0", "?c"))),
```

Comment:
The query plans are nicer if this is only a single filter with `&&` or `NOT IN()`.

Reply:
(Currently the `(?x != <a>) && (?x != <b>)` is more efficient than `?x NOT IN(<a>, <b>)`.